# Project Name - Automated Sales Reporting & Analystics System
# Section - Data Validation
2. Missing Value Validation
3. Duplicate Validation 
4. Uniqueness Validation
5. Categorical Value Validation
6. Numerical Range Validation
7. Format Validation
8. Outlier Validation
9. Data Accuracy Validation
10. Data Quality Validation

In [52]:
# Libraries 
import pandas as pd
import seaborn as sns
import statistics as st
import numpy as np
from IPython.display import display

In [53]:
# load row data into pandas
dt=pd.read_csv('C:\\Users\lavir\OneDrive\Pictures\Documents\Automated_Sales_Reporting\dataset\Food_Delivery_Operations_Analytics_messy_50k.csv')

In [54]:
dt.tail(10)

,Order_ID,Customer_Age,Customer_Type,Restaurant_Type,Cuisine_Type,Delivery_Distance_KM,Delivery_Time_Min,Order_Value_USD,Item_Count,Weather_Condition,...,Customer_Rating,Complaint_Flag,Refund_Flag,Revenue_USD,Profit_USD,City,Month,Quarter,Demand_Score,Churn_Risk
49990,ORD0072733,67,Premium,Cloud Kitchen,Chinese,16.92,68,108.29,6,Cloudy,...,3.6,0,0,17.4,186.5,London,4,2,48.22,96.71
49991,ORD0105016,32,New,Restaurant,Italian,17.01,91,$265.49,4,Cloudy,...,4.8,0,NaN,67.08,101.75,Mumbai,7,3,42.41,46.87
49992,ORD0096588,18,Returning,Cafe,Indian,22.15,44,182.22,2,Stormy,...,1.4,0,0,307.02,185.44,Singapore,2,1,30.26,29.88
49993,ORD0136949,40,New,Cafe,Indian,5.52,79,124.45,4,Cloudy,...,1.2,False,0,125.75,40.07,London,6,2,48.39,47.22
49994,ORD0069863,55,Returning,Restaurant,-,10.56,35,271.64,12,Cloudy,...,4.7,0,0,461.59,-2.71,Singapore,12,4,2.64,21.54
49995,ORD0117771,45,Returning,Cafe,Italian,1.82,82,140.8,13,Rainy,...,,No,0,252.7,-7.87,Singapore,11,4,70.41,53.31
49996,ORD0076451,56,New,Restaurant,Chinese,6.04,54,72.93,2,Sunny,...,3.9,1,unknown,295.18,4.73 USD,Singapore,6,2,96.25,7.7
49997,ORD0040166,53,Premium,Cloud Kitchen,Indian,2.29,75,297.22,12,Sunny,...,5.0,0,0,315.44,162.29,New York,12,4,6.61,15.37
49998,ORD0107152,33,Premium,Cloud Kitchen,Chinese,8.39,71,28.01,2,Stormy,...,4.0,0,0,157.28,44.26,New York,4,2,18.66,67.75
49999,ORD0020512,68,Premium,Fast Food,Mexican,12.67,51,266.37,13,Sunny,...,2.6,1,0,402.06,106.43,Singapore,7,3,92.8,15.63


# My Basic Understanding About Columns 
1. Categorical Data Columns – Order_ID, Customer_Type, Restaurant_Type, Cuisine_Type, Weather_Condition, Complaint_Flag, Refund_Flag, City
2. Numerical Data Columns – Customer_Age, Delivery_Distance_KM, Delivery_Time_Min, Order_Value_USD, Item_Count, Customer_Rating, Revenue_USD, Profit_USD, Month, Quarter, Demand_Score, Churn_Risk
3. Fill/Replace:
Cuisine_Type → replace blanks or - with "Unknown"
Complaint_Flag → replace blanks with "No" and standardize to 0/1
Refund_Flag → replace blanks with "No" and standardize to 0/1
Customer_Rating → fill blanks with median value
4. Remove/Drop:
Order_ID → drop rows if missing (ID must be unique)
5. Leave As Is / Optional:
Profit_USD → allow negatives (valid business losses), clean "USD" text only
Other numerical columns (Customer_Age, Delivery_Distance_KM, Delivery_Time_Min, Item_Count, Revenue_USD, Demand_Score, Churn_Risk) → drop rows if blanks are rare, no imputation needed
Other categorical columns (Customer_Type, Restaurant_Type, Weather_Condition, City) → blanks can stay or be replaced with "Unknown" if desired



In [55]:
# Missing Value Validation for each columns 
# And also clean data percentage and missing data percentage 
def missing_value(df):
    print("=======Complete Missing Values Report=======")
    report = pd.DataFrame({
        'Missing_Count': df.isnull().sum(),
        'Missing_%': (df.isnull().sum() / len(df)) * 100,
        'Clean_%': 100 - ((df.isnull().sum() / len(df)) * 100)
    })
    return report
a=missing_value(dt)
print(a)
    

=======Complete Missing Values Report=======
                      Missing_Count  Missing_%  Clean_%
Order_ID                          0      0.000  100.000
Customer_Age                    788      1.576   98.424
Customer_Type                     0      0.000  100.000
Restaurant_Type                   0      0.000  100.000
Cuisine_Type                    752      1.504   98.496
Delivery_Distance_KM           1285      2.570   97.430
Delivery_Time_Min               756      1.512   98.488
Order_Value_USD                1278      2.556   97.444
Item_Count                        0      0.000  100.000
Weather_Condition               739      1.478   98.522
Traffic_Level                   748      1.496   98.504
Customer_Rating                1263      2.526   97.474
Complaint_Flag                    0      0.000  100.000
Refund_Flag                     777      1.554   98.446
Revenue_USD                       0      0.000  100.000
Profit_USD                        0      0.000  100.000
Cit

In [56]:
# Duplicate And Uniquess Validation For Order_id
def duplicated_data(data,col_name):
    print(f"======Duplicate Validation {col_name}===== ")
    duplicated_data=data[col_name].duplicated().sum()
    return (duplicated_data)

result=duplicated_data(dt,'Order_ID')
print(result)

======Duplicate Validation Order_ID===== 
757


In [57]:
# Remove Duplicates Value From Order_id
dt1 = dt.drop_duplicates() # Remove Duplicates row of every columns 


In [58]:
# Now, Check Duplicates
a=dt1.duplicated().sum() 
if a ==0: 
 print('========Duplicate value remove sucessfully======') 
 print(a)
else:
  print('Still,Pending duplicates value')

========Duplicate value remove sucessfully======
0


In [59]:
# Remove missing values from every columns
dt2=dt1.dropna()
print('After removed missing values',dt2.shape)


After removed missing values (39720, 21)


# After Remove Dataset Summary

- **Before Cleaning:** 50,000 rows × 21 columns  
- **After Cleaning:**  39720 rows × 21 columns  

**Change:** Removed 10,280 rows with missing values.  
**Result:** Cleaner dataset, same schema, ready for analysis.


In [60]:
# Checking unique values in every columns
for col in dt2.columns:
    print(f"\n===== {col} =====")
    print("Unique Values:", dt2[col].unique())


===== Order_ID =====
Unique Values: ['ORD0140396' 'ORD0111309' 'ORD0053947' ... 'ORD0040166' 'ORD0107152'
 'ORD0020512']

===== Customer_Age =====
Unique Values: ['47' '28' '44' '70' '45' '64' '43' '65' '59' '26' '63' '23' '66' '32'
 '74' '72' '57' '38' '71' '35' '24' '29' '40' '55' '50' '22' '41' '51'
 '39' '48' '69' '52' '31' '62' '25' '53' '56' '36' '58' '19' '30' '33'
 '37' '42' '21' '67' '61' '18' '999' '-5' '73' '54' '46' '27' '49' '20'
 '68' '-' '34' '60' ' ' 'unknown' '0' '150']

===== Customer_Type =====
Unique Values: ['Premium' 'New' 'Returning' '  Premium  ' 'PREMIUM' '  New  ' 'returning'
 'Returning ' 'new' 'NEW' 'RETURNING' 'Premium ' 'premium' 'New '
 '  Returning  ']

===== Restaurant_Type =====
Unique Values: ['Cafe' 'Cloud Kitchen' 'Restaurant' 'Fast Food' 'Cloud Kitchen '
 '  Cloud Kitchen  ' 'cloud kitchen' 'RESTAURANT' '  Fast Food  '
 'fast food' 'Fast Food ' 'CLOUD KITCHEN' 'Restaurant ' 'FAST FOOD'
 '  Cafe  ' 'CAFE' 'restaurant' 'cafe' '  Restaurant  ' 'Cafe 

In [61]:
# Identify wrong data in dt
df=dt2[~dt2['Order_ID'].str.match(r'^ORD\d{7}$')]
df['Order_ID']

5        ORD0142185_DUP
136      ORD0142907_DUP
200      ORD0048728_DUP
207         ORD0022916 
224      ORD0032208_DUP
              ...      
49850       ord-0093276
49945       ord-0016616
49955    ORD0061689_DUP
49972    ORD0139717_DUP
49987    ORD0135388_DUP
Name: Order_ID, Length: 872, dtype: object

In [62]:
# Replace '_DUP' into '' 
dt2['Order_ID']=dt2['Order_ID'].str.replace('_DUP','')
print(dt2['Order_ID'].unique()[:10])  # first 10 unique IDs after cleaning

['ORD0140396' 'ORD0111309' 'ORD0053947' 'ORD0018219' 'ORD0142185'
 'ORD0051220' 'ORD0132839' 'ORD0002114' 'ORD0136381' 'ORD0016784']


C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\4017055438.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Order_ID']=dt2['Order_ID'].str.replace('_DUP','')


In [63]:
# Replace '-' into '' 
dt2['Order_ID']=dt2['Order_ID'].str.replace('-','')
# Remove extra space 
dt2['Order_ID']=dt2['Order_ID'].str.strip()
# all lower case into upper case
dt2['Order_ID']=dt2['Order_ID'].str.upper()


C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\1250939447.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Order_ID']=dt2['Order_ID'].str.replace('-','')
C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\1250939447.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Order_ID']=dt2['Order_ID'].str.strip()
C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\1250939447.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_

# Data Cleaning - Customer_Age

In [64]:
# replace all -,'',unknown with NaN
dt2['Customer_Age'].replace(["-","","unknown"],np.nan)


1        47
2        28
3        44
4        70
5        45
         ..
49995    45
49996    56
49997    53
49998    33
49999    68
Name: Customer_Age, Length: 39720, dtype: object

In [65]:
# object into numeric column
dt2['Customer_Age']=pd.to_numeric(dt2['Customer_Age'],errors='coerce')


C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\849169800.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Customer_Age']=pd.to_numeric(dt2['Customer_Age'],errors='coerce')


In [66]:
# Now, NaN fill by median 
dt2['Customer_Age'].fillna(dt2['Customer_Age'].median,inplace=True)

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\1816163417.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '<bound method Series.median of 1        47.0
2        28.0
3        44.0
4        70.0
5        45.0
         ... 
49995    45.0
49996    56.0
49997    53.0
49998    33.0
49999    68.0
Name: Customer_Age, Length: 39720, dtype: float64>' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  dt2['Customer_Age'].fillna(dt2['Customer_Age'].median,inplace=True)
C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\1816163417.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Customer_Age'].fillna(dt2['Customer_Age'].median,inplace=True)


In [67]:
# after  fill NaN values with the median so, verify if any NaN or null values are still left using these checks:
dt2['Customer_Age'].notna().sum()

np.int64(39720)

Data Cleaning In Customer_Type

In [68]:
# When I checked, I found that the values in Customer_Type were in mixed case so, convert into title case (first letter capital)
# Remove extra spaces
dt2["Customer_Type"] = dt2["Customer_Type"].str.strip()

# Convert to Title Case
dt2["Customer_Type"] = dt2["Customer_Type"].str.title()

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\2613730465.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2["Customer_Type"] = dt2["Customer_Type"].str.strip()
C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\2613730465.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2["Customer_Type"] = dt2["Customer_Type"].str.title()


In [69]:
'''1. Categorical Data Columns – Order_ID, Customer_Type, Restaurant_Type, Cuisine_Type,   Weather_Condition, Complaint_Flag, Refund_Flag, City
2. Numerical Data Columns – Customer_Age, Delivery_Distance_KM, Delivery_Time_Min, 
Order_Value_USD, Item_Count, Customer_Rating, Revenue_USD, Profit_USD, Month, Quarter,
Demand_Score, Churn_Risk'''

'1. Categorical Data Columns – Order_ID, Customer_Type, Restaurant_Type, Cuisine_Type,   Weather_Condition, Complaint_Flag, Refund_Flag, City\n2. Numerical Data Columns – Customer_Age, Delivery_Distance_KM, Delivery_Time_Min, \nOrder_Value_USD, Item_Count, Customer_Rating, Revenue_USD, Profit_USD, Month, Quarter,\nDemand_Score, Churn_Risk'

data cleaning in restaurant_type

In [70]:
# remove extra spaces in Restaurant_Type
dt2['Restaurant_Type']=dt2['Restaurant_Type'].str.strip()
# mixed cases convert into title case
dt2['Restaurant_Type']=dt2['Restaurant_Type'].str.title()

# After completing the cleaning, now checking the unique values in  Restaurant_Type
display(dt2['Restaurant_Type'].unique())

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\4177498520.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Restaurant_Type']=dt2['Restaurant_Type'].str.strip()
C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\4177498520.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Restaurant_Type']=dt2['Restaurant_Type'].str.title()


array(['Cafe', 'Cloud Kitchen', 'Restaurant', 'Fast Food'], dtype=object)

data cleaning in Cuisine_Type,   Weather_Condition, Complaint_Flag, Refund_Flag, City

In [71]:
dt2['Cuisine_Type']=dt2['Cuisine_Type'].str.strip().str.title() # Remove extra space and convert into title'
dt2['Weather_Condition']=dt2['Weather_Condition'].str.strip().str.title() # Remove extra space and convert into title  

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\1880648950.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Cuisine_Type']=dt2['Cuisine_Type'].str.strip().str.title() # Remove extra space and convert into title'
C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\1880648950.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Weather_Condition']=dt2['Weather_Condition'].str.strip().str.title() # Remove extra space and convert into title


In [72]:
dt2['Cuisine_Type']=dt2['Cuisine_Type'].astype(str).replace(['','-','Nan','null','Null','N/A'],"Unknown") # replace all '-','Nan',etc into "Unknown"
dt2['Cuisine_Type'].unique() 

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\320559172.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Cuisine_Type']=dt2['Cuisine_Type'].astype(str).replace(['','-','Nan','null','Null','N/A'],"Unknown") # replace all '-','Nan',etc into "Unknown"


array(['Chinese', 'Indian', 'Italian', 'Mexican', 'Unknown'], dtype=object)

In [73]:
dt2['Weather_Condition']=dt2['Weather_Condition'].astype(str).replace(['','-','Nan','null','Null',
'N/A'],"Unknown") # replace all '-','Nan',etc into "Unknown"
dt2['Weather_Condition'].unique() 

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\2804236183.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Weather_Condition']=dt2['Weather_Condition'].astype(str).replace(['','-','Nan','null','Null',


array(['Rainy', 'Cloudy', 'Stormy', 'Sunny', 'Unknown'], dtype=object)

In [74]:
# For Complaint_Flag
# Replace - ("0","False","No") with 0 and ("Yes","True","1") with 1
dt2['Complaint_Flag']=dt2['Complaint_Flag'].replace({"False":0,"0":0,"No":0,"Yes":1,"True":1,"1":1}).astype(int)
dt2['Complaint_Flag'].unique() # after dtype conversion and replaced values 

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\4118277720.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dt2['Complaint_Flag']=dt2['Complaint_Flag'].replace({"False":0,"0":0,"No":0,"Yes":1,"True":1,"1":1}).astype(int)
C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\4118277720.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Complaint_Flag']=dt2['Complaint_Flag'].replace({"False":0,"0":0,"No":0,"Yes":1,"True":1,"1":1}).astype(int)


array([0, 1])

In [75]:
# For Refund_Flag
dt2['Refund_Flag']=dt2['Refund_Flag'].replace({'False':0,'No':0,' ':np.nan,'-':np.nan,'Yes':1, 'unknown':np.nan,'True':1}).fillna(0).astype(int) # Standardized boolean values (False/No → 0, True/Yes → 1), converted blank/- values to NaN, and filled missing values with 0.
dt2['Refund_Flag'].unique() # after complete all process


C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\1907074278.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['Refund_Flag']=dt2['Refund_Flag'].replace({'False':0,'No':0,' ':np.nan,'-':np.nan,'Yes':1, 'unknown':np.nan,'True':1}).fillna(0).astype(int) # Standardized boolean values (False/No → 0, True/Yes → 1), converted blank/- values to NaN, and filled missing values with 0.


array([0, 1])

In [76]:
# For City columns 
dt2['City']=dt2['City'].str.strip() # Remove extra spacing from city column values

dt2['City']=dt2['City'].str.title() # convert inot title case


C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\3479896230.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['City']=dt2['City'].str.strip() # Remove extra spacing from city column values
C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\3479896230.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['City']=dt2['City'].str.title() # convert inot title case


In [77]:
# Replace '-','','Null',etc into 'Unknown' for city column 
dt2['City']=dt2['City'].replace(['','-','Null','Nan','N/A'],"Unknown").astype(str)

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\2126245499.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['City']=dt2['City'].replace(['','-','Null','Nan','N/A'],"Unknown").astype(str)


In [78]:
 # City Name Standardization and Validation for city column
dt2['City']=dt2['City'].replace({'Sydney, Au':'Sydney', 'Sg': 'Singapore', 'London, Uk':'London', 'New York City': 'New York', 'Ny':'New York','Newyork':'New York'}).astype(str)
dt2["City"].unique()

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\3557137133.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2['City']=dt2['City'].replace({'Sydney, Au':'Sydney', 'Sg': 'Singapore', 'London, Uk':'London', 'New York City': 'New York', 'Ny':'New York','Newyork':'New York'}).astype(str)


array(['Sydney', 'New York', 'Mumbai', 'London', 'Singapore', 'Unknown',
       'Bombay'], dtype=object)

Data cleaning & validation for numerical columns - Customer_Age, Delivery_Distance_KM, Delivery_Time_Min, Order_Value_USD, Item_Count, Customer_Rating, Revenue_USD, Profit_USD, Month, Quarter, Demand_Score, Churn_Risk


In [79]:
dt2["Delivery_Distance_KM"]=dt2["Delivery_Distance_KM"].str.replace(" km", "",regex=True).replace({"-":np.nan,"unknown":np.nan,"Unknown":np.nan,"":np.nan," ":np.nan}).astype(float)

C:\Users\lavir\AppData\Local\Temp\ipykernel_38932\923378886.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt2["Delivery_Distance_KM"]=dt2["Delivery_Distance_KM"].str.replace(" km", "",regex=True).replace({"-":np.nan,"unknown":np.nan,"Unknown":np.nan,"":np.nan," ":np.nan}).astype(float)


In [80]:
dt2 = dt2.dropna(subset=["Delivery_Distance_KM"]) # Remove all missing values from 'Delivery_Distance_KM' 
dt2['Delivery_Distance_KM'].unique()

array([  9.7 ,   2.97,   5.82, ...,  -8.83,  -2.54, -14.67], shape=(2609,))

In [81]:
# For Delivery_Time_Min

# remove suffix 'm','unknown' 
dt2['Delivery_Time_Min'] = (
    dt2['Delivery_Time_Min']
    .replace([' ','-','nan','unknown','Unknown'], np.nan)   # messy values ko NaN
    .astype(str)                                 # sabko string banao
    .str.replace(r"(min|in)$", "", regex=True)   # suffix 'min' ya 'in' hatao
)


In [82]:
dt2["Delivery_Time_Min"] = dt2["Delivery_Time_Min"].replace("nan", np.nan) # Again 'nan' value convert into NaN
dt2=dt2.dropna(subset=['Delivery_Time_Min']) # Now, Remove all missing values or NaN inside the columns
display(dt2['Delivery_Time_Min'].unique()) # after complete all process, again check the unquies values 

array(['71', '107', '43', '45', '38', '68', '52', '39', '55', '86', '85',
       '111', '118', '51', '48', '53', '93', '73', '33', '34', '64', '98',
       '28', '83', '82', '13', '77', '92', '103', '59', '91', '65', '67',
       '116', '11', '106', '22', '17', '54', '26', '63', '40', '119',
       '115', '78', '79', '50', '88', '20', '100', '62', '32', '81', '94',
       '82 ', '36', '105', '117', '113', '60', '47', '97', '-16', '75',
       '84', '29', '109', '37', '15', '99', '69', '25', '104', '44', '21',
       '10', '-29', '101', '31', '24', '58', '95', '18', '23', '114',
       '14', '76', '30', '112', '72', '80', '35', '96', '-59', '90', '49',
       '19', '70', '87', '110', '66', '102', '108', '57', '63 ', '41',
       '61', '89 ', '56', '50 ', '83 ', '27', '46', '16', '42', '55 ',
       '12', '60 ', '109 ', '84 ', '89', '74', '-28', '76 ', '-23', '99 ',
       '58 ', '-49', '118 ', '37 ', '45 ', '103 ', '86 ', '48 ', '-19',
       '13 ', '73 ', '87 ', '64 ', '62 ', '93 ', '4

In [83]:
# For Order_Value_USD
dt2['Order_Value_USD']=(dt2['Order_Value_USD']
                        .replace([" ","","-","unknown","Unknown"],np.nan) 
                        .astype(str).replace(r"(USD|usd)$", "",regex=True)
                        )

In [84]:
dt2=dt2.dropna(subset=['Order_Value_USD']) # Remove missing values 
dt2['Order_Value_USD'].unique()  

array(['16.19', '289.91', '87.03', ..., '182.22', '124.45', '72.93'],
      shape=(22325,), dtype=object)

In [85]:
# For Item_Count
dt2['Item_Count'].unique() 
# Item_Count column is already clean, with no null or missing values.

array([13,  1,  4, 12,  9, 14, 10,  6,  2,  3, 11,  5,  7,  8])

In [86]:
# For Customer_Rating 
dt2['Customer_Rating']=dt2['Customer_Rating'].replace([' ','','-','unknown'],np.nan).astype(float)
dt2=dt2.dropna(subset=['Customer_Rating'])
dt2['Customer_Rating'].unique()

array([ 4.8, 10. ,  1.3,  4. ,  2.9,  3.2,  3.6,  2.5,  1.2,  4.2,  4.5,
        2.7,  3.8,  2.2,  4.4,  3.4,  4.1,  3.9,  4.6,  5. ,  2.6,  2.3,
        2.4,  2.1,  4.7,  4.9,  1.8,  3.5,  2. ,  1.1,  3.7,  3. ,  3.1,
        1. ,  2.8,  1.6,  1.9,  1.7,  1.5,  1.4,  4.3,  3.3, -1. ,  0. ,
        5.5,  6. ])

In [87]:
dt2.loc[~dt2['Customer_Rating'].between(1, 5), 'Customer_Rating'] = np.nan # check all customer_rating if have any value <=10 so, replace with median
dt2['Customer_Rating']=dt2['Customer_Rating'].fillna(dt2['Customer_Rating'].median()) # Replace with median

In [88]:
# For Revenue_USD Column
# Step 1: Replace blanks and unknowns with NaN
dt2['Revenue_USD'] = dt2['Revenue_USD'].replace(['','unknown','Unknown','Uknown','-',' '], np.nan)

# Step 2: Remove dollar sign and 'USD' text
dt2['Revenue_USD'] = dt2['Revenue_USD'].replace(r'\$', '', regex=True)
dt2['Revenue_USD'] = dt2['Revenue_USD'].replace(r'USD', '', regex=True)

# Step 3: Strip extra spaces
dt2['Revenue_USD'] = dt2['Revenue_USD'].str.strip()



In [89]:
# convert Revenue_USD column into float
dt2['Revenue_USD']=dt2['Revenue_USD'].astype(float)

In [90]:
# For Profit_USD column
dt2['Profit_USD']=dt2['Profit_USD'].replace(['',' ','-','unknown','Unknown'],np.nan) # replace blanks,'-','Unknown' with NaN
# Remove $ and USD sign
dt2['Profit_USD']=dt2['Profit_USD'].replace(r'\$','',regex=True)
dt2['Profit_USD']=dt2['Profit_USD'].replace(r'USD','',regex=True)

# convert into float and positive
dt2['Profit_USD']=dt2['Profit_USD'].astype(float)
dt2['Profit_USD']

1         23.90
2         55.43
3         34.38
4         -7.46
5         78.70
          ...  
49994     -2.71
49996      4.73
49997    162.29
49998     44.26
49999    106.43
Name: Profit_USD, Length: 38610, dtype: float64

['10', '9', '7', '1', '4', '12', '6', '3', 'February', '8', '5',
       '11', 'April', '2', 'Jan', 'December', 'Jun', 'Jul', 'January',
       'March', 'July', 'May', 'Aug', 'Mar', 'Apr', 'June', 'Dec', 'Oct',
       'November', 'September', 'Sep', 'Nov', 'August', 'Feb', 'October'],

In [91]:
# For month columns
dt2['Month']=dt2['Month'].replace({'1':'January','2':'February','3':'March','4':'April','5':'May','6':'June','7':'July','8':'augest','9':'September','10':'October','11':'November','12':'December','Jan':'January','Jun':'June','Jul':'july','Nov':'November','Feb':'February','Aug':'Augest', 'Mar':'March', 'Apr':'April', 'Dec':'December', 'Oct':'October', 'Sep':'September','August':'Augest'}).astype(str) # replace all january to december
dt2['Month']=dt2['Month'].str.title() # convert into title case


In [92]:
dt2['Month'].unique() 

array(['October', 'September', 'July', 'January', 'April', 'December',
       'June', 'March', 'February', 'Augest', 'May', 'November'],
      dtype=object)

In [93]:
# For Quarter column
dt2['Quarter'].unique() 
# No requried any cleaing, validation because its all ok 

array([4, 3, 1, 2])

In [94]:
# For Demand_Score Columns
dt2=dt2.drop(columns=['Demand_Score'])


In [95]:
# For Churn_Risk Column
dt2['Churn_Risk']=dt2['Churn_Risk'].replace(['',' ','-','unknown','Unknown'],np.nan).astype(float) # replace all '-','unknown' with NaN


In [96]:
# For Traffic_Level

dt2['Traffic_Level']=dt2['Traffic_Level'].str.strip() # remove extra spaces i
dt2['Traffic_Level']=dt2['Traffic_Level'].str.lower()  # convert into lower cases 
dt2['Traffic_Level']=dt2['Traffic_Level'].replace(['','-','null','nan','n/a','unknown',],'Unknown') # Replace '-','','unknown','NaN','N/A' and 'null' into 'Unknown' 
dt2['Traffic_Level']=dt2['Traffic_Level'].str.title() # convert into title case




In [97]:
dt2['Traffic_Level'].unique()


array(['High', 'Medium', 'Low', 'Unknown'], dtype=object)

In [98]:
# Customer_Age, Delivery_Time_Min, Order_Value_USD
# Customer_Age → convert safely to int
dt2['Customer_Age'] = pd.to_numeric(dt2['Customer_Age'], errors='coerce').astype('Int64')

# Delivery_Time_Min → convert safely to int
dt2['Delivery_Time_Min'] = pd.to_numeric(dt2['Delivery_Time_Min'], errors='coerce').astype('Int64')

# Order_Value_USD → convert safely to float
dt2['Order_Value_USD'] = pd.to_numeric(dt2['Order_Value_USD'], errors='coerce').astype(float)

In [99]:
#check duplicates value in Order_ID
dt2['Order_ID'].duplicated().sum() 


np.int64(563)

In [100]:
# Remove duplicates values in Order_ID
dt2.drop_duplicates(subset='Order_ID',inplace=True)

In [101]:
# Converted binary 0/1 values in Refund_Flag and Complaint_Flag into meaningful categorical labels.
dt2['Refund_Flag']=dt2['Refund_Flag'].map({

    0:"No Refund",
    1: "Refund"
})   # For Refund_Flag

dt2['Complaint_Flag']=dt2['Complaint_Flag'].map({
    0:"No Complaint",
    1:"Complaint"
})


In [102]:
# check binary into Refund_Flag and Complaint_Flag
dt2[['Refund_Flag','Complaint_Flag']]

,Refund_Flag,Complaint_Flag
1,No Refund,No Complaint
2,No Refund,No Complaint
3,No Refund,No Complaint
4,No Refund,No Complaint
5,No Refund,No Complaint
...,...,...
49994,No Refund,No Complaint
49996,No Refund,Complaint
49997,No Refund,No Complaint
49998,No Refund,No Complaint


In [104]:
dt2.dtypes

Order_ID                 object
Customer_Age              Int64
Customer_Type            object
Restaurant_Type          object
Cuisine_Type             object
Delivery_Distance_KM    float64
Delivery_Time_Min         Int64
Order_Value_USD         float64
Item_Count                int64
Weather_Condition        object
Traffic_Level            object
Customer_Rating         float64
Complaint_Flag           object
Refund_Flag              object
Revenue_USD             float64
Profit_USD              float64
City                     object
Month                    object
Quarter                   int64
Churn_Risk              float64
dtype: object

In [105]:
# save changes file 
dt2.to_excel('C:\\Users\lavir\OneDrive\Pictures\Documents\Automated_Sales_Reporting\dataset\clean_dataset.xlsx',index=False,engine='openpyxl')